src/main/java/net/minecraftforge/fml/common/versioning/DependencyParser.java

```diff
@@ -0,0 +1,235 @@
+            {
+                throw new DependencyParserException(dep, "Cannot have an \"all\" (*) relationship with anything except pure *");
+            }
+            else if (targetIsBounded)
+            {
+                throw new DependencyParserException(dep, "You cannot have a versioned dependency on everything (*)");
+            }
+        }
+
+        Side depSide = null;
+        String depOrder = null;
+        boolean depRequired = false;
+
+        for (String instruction : instructions)
+        {
+            if ("client".equals(instruction))
+            {
+                if (depSide != null)
+                {
+                    throw new DependencyParserException(dep, "Up to one side (client or server) can be specified.");
+                }
+                depSide = Side.CLIENT;
+            }
+            else if ("server".equals(instruction))
+            {
+                if (depSide != null)
+                {
+                    throw new DependencyParserException(dep, "Up to one side (client or server) can be specified.");
+                }
+                depSide = Side.SERVER;
+            }
+            else if ("required".equals(instruction))
+            {
+                if (depRequired)
+                {
+                    throw new DependencyParserException(dep, "'required' can only be specified once.");
+                }
+                if (targetIsAll)
+                {
+                    throw new DependencyParserException(dep, "You can't 'require' everything (*)");
+                }
+                depRequired = true;
+            }
+            else if ("before".equals(instruction) || "after".equals(instruction))
+            {
+                if (depOrder != null)
+                {
+                    throw new DependencyParserException(dep, "'before' or 'after' can only be specified once.");
+                }
+                depOrder = instruction;
+            }
+            else
+            {
+                throw new DependencyParserException(dep, String.format("Found invalid instruction '%s'. Only %s are allowed.", instruction, DEPENDENCY_INSTRUCTIONS));
+            }
+        }
+
+        if (depSide != null && depSide != this.side)
+        {
+            return;
+        }
+
+        ArtifactVersion artifactVersion;
+        try
+        {
+            artifactVersion = VersionParser.parseVersionReference(target);
+        }
+        catch (RuntimeException e)
+        {
+            throw new DependencyParserException(dep, "Could not parse version string.", e);
+        }
+
+        // TODO: enable this in 1.13
+//        if (!targetIsAll)
+//        {
+//            String modId = artifactVersion.getLabel();
+//            sanityCheckModId(dep, modId);
+//        }
+
+        if (!depRequired && depOrder == null && !targetIsBounded)
+        {
+            throw new DependencyParserException(dep, "Soft dependencies must have a version restriction specified.");
```

Comment:
Now that the other thing is out of the way: I am not quite sure what the reasoning behind this restriction would be. To give an example: I might write a Kotlin mod and hence depend on Forgelin. However, I could technically not be using anything special from the Kotlin standard library (except for its auto-generated null-checking stuff, but I'm going to assume that they are pretty much set in stone), so I don't require any particular version of Forgelin/Kotlin, either. Of course I could just specify a version range without an upper bound, but I don't really see a reason for not allowing version-less soft dependencies.

Reply:
If you need forgelin of any version, you say `"required:forgelin;"`.
If you have a soft dependency on forgelin of any version, you don't need to write any dependency restriction at all, it would do nothing.

Reply:
Derp² I shouldn't be looking at this right after getting out of bed xD